src/Shared/VisualStudioLocationHelper.cs

```diff
@@ -0,0 +1,123 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Runtime.InteropServices;
+using Microsoft.VisualStudio.Setup.Configuration;
+
+namespace Microsoft.Build.Shared
+{
+    internal class VisualStudioInstance
+    {
+        /// <summary>
+        /// Version of the Visual Studio Instance
+        /// </summary>
+        internal Version Version { get; }
+
+        /// <summary>
+        /// Path to the Visual Studio installation
+        /// </summary>
+        internal string Path { get; }
+
+        /// <summary>
+        /// Full name of the Visual Studio instance with SKU name
+        /// </summary>
+        internal string Name { get; }
+
+        internal VisualStudioInstance(string name, string path, Version version)
+        {
+            Name = name;
+            Path = path;
+            Version = version;
+        }
+    }
+
+    internal class VisualStudioLocationHelper
+    {
+        private const int REGDB_E_CLASSNOTREG = unchecked((int)0x80040154);
+
+        /// <summary>
+        /// Query the Visual Studio setup API to get instances of Visual Studio installed
+        /// on the machine. Will not include anything before Visual Studio "15".
+        /// </summary>
+        /// <returns>Enumerable list of Visual Studio instances</returns>
+        internal static IEnumerable<VisualStudioInstance> GetInstances()
```

Comment:
This generally seems a bit more complicated than I'd hoped, and it isn't super well commented. Is this as easy as it gets?